### PR TITLE
feat(parser, compiler): add parsing a type reference from a string

### LINKED ISSUE #715: add parsing support for `Type`
Similar to #681, `@field` directive can be provided with `Type` information in a string form, which will then need to be parsed before being used in the schema. It seems like it does not need to be validated specifically (but we should do it anyways)

To make this change, we need to:

1. add `parse_type` method to `apollo-parser`
2. add a `cst::Type` specific method for `SyntaxTree` implementation to return the CST node, similar to doc.document(), and field_set.field_set()
3. create a `parse_type` method in `apollo-compiler`
4. add `Type` struct to `apollo-compiler`'s schema that would also be able to include `BuildErrors`
5. add a `parse` and `validate` methods to the `Type` struct impl

### PR DESCRIPTION
**Pull request by lrlna**
To parse standalone source text like `[Type]!`, for example for the federation directive [`@fields(type:)`](https://specs.apollo.dev/join/v0.3/#@field). 

closes #715

### CODE REVIEW

File: crates/apollo-compiler/src/parser.rs [hunk 1]
```diff
@@ -234,6 +236,35 @@ impl Parser {
...
 
+    /// Parse the given source of a field type.
+    ///
+    /// `path` is the filesystem path (or arbitrary string) used in diagnostics
+    /// to identify this source file to users.
+    pub fn parse_field_type(
+        &mut self,
+        source_text: impl Into<String>,
+        path: impl AsRef<Path>,
+    ) -> Result<(schema::FieldType, Diagnostics), Diagnostics> {
```
Comment (lrlna) [inline crates/apollo-compiler/src/parser.rs:247]: Returning a `Result` here for field types which are not syntactically valid (and therefore will not have a FieldType), but will have syntax diagnostics, e.g. `[]` as a field type.

Comment (goto-bus-stop) [reply crates/apollo-compiler/src/parser.rs:247]: `Diagnostics` can only have syntax errors or limit errors here, so perhaps we can just have `Result<FieldType, Diagnostics>` and always return `Err(errors)` if there are any?

**Comment by goto-bus-stop** [comment]
Also Also Also, should this just be parsing into a `Type` directly instead of into a `FieldType` wrapper? A type reference doesn't *have* to be in a field, field/directive argument types are the same.

**Comment by SimonSapin** [comment]
Agree on both points: this can use `ast::Type` without introducing a `FieldType` struct, and the `Ok` result doesn’t need to contain diagnostics (or conversely, return `Err` if any diagnostic)

**Comment by goto-bus-stop** [comment]
@SimonSapin Ready for review again I think!

File: crates/apollo-parser/src/parser/mod.rs [hunk 2]
```diff
@@ -146,12 +146,15 @@ impl<'a> Parser<'a> {
...
-            syntax_tree::SyntaxTreeWrapper::FieldSet(_) => {
+            syntax_tree::SyntaxTreeWrapper::Type(_)
+            | syntax_tree::SyntaxTreeWrapper::FieldSet(_) => {
                 unreachable!("parse constructor can only construct a document")
             }
         }
     }
 
+    /// Parse selection set tokens. Useful for specifically parsing selections
+    /// sets which are part of specific directives, like `@requires`.
```
Comment (SimonSapin) [inline crates/apollo-parser/src/parser/mod.rs:157]:
```suggestion
    /// Parse a selection set with optional outer braces.
    /// This is the expected format of the string value of the `fields` argument of some directives
    /// like [`@requires`](https://www.apollographql.com/docs/federation/federated-types/federated-directives/#requires).
```

Comment (SimonSapin) [reply crates/apollo-parser/src/parser/mod.rs]: Braces? Brackets?

Comment (goto-bus-stop) [reply crates/apollo-parser/src/parser/mod.rs]: IMO braces means `{}`, and brackets can mean that or one of the other pairs

**Pull request merged**